### PR TITLE
Issue 2933: (SegmentStore) testEndToEndWithFencing fixes

### LINKED ISSUE #2933: Sporadic Failure in HDFSIntegrationTest.testEndToEndWithFencing
**Problem description**
The exception causing the test failure is:
```
java.util.concurrent.TimeoutException
    at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1771)
    at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
    at io.pravega.segmentstore.server.store.StreamSegmentStoreTestBase.testEndToEndWithFencing(StreamSegmentStoreTestBase.java:296)
```
Following the flow of the test, it can be seen that everything goes as expected until it executes `waitForSegmentsInStorage`, which is the point where the timeout exception arises.

When the test gets into this method, we start observing `BadOffsetException` messages related with a particular transaction segment:

```
2018-09-18 06:01:51,980 INFO  [storage-io-2] writer.SegmentAggregator (SegmentAggregator.java:setState(1549)) - StorageWriter[3-556]: State changed from Writing to ReconciliationNeeded.
2018-09-18 06:01:51,980 ERROR [storage-io-2] writer.StorageWriter (StorageWriter.java:logError(457)) - StorageWriter[3]: Iteration[2].Error.
io.pravega.segmentstore.contracts.BadOffsetException: [Segment 'Segment_0#transaction.bf2cc480df7546f39eddaacf227a1a8a$offset.0'] Bad Offset. Expected 1700, given 0.
    at io.pravega.segmentstore.storage.impl.hdfs.HDFSStorage.write(HDFSStorage.java:360)
    at io.pravega.segmentstore.storage.rolling.RollingStorage.write(RollingStorage.java:322)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$write$2(AsyncStorageWrapper.java:100)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$runAsync$15(AsyncStorageWrapper.java:206)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.execute(AsyncStorageWrapper.java:221)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$supplyAsync$13(AsyncStorageWrapper.java:181)
```
As well as `IllegalArgumentException` in storage writer related to offsets:
```
2018-09-18 06:01:52,997 INFO  [IPC Server handler 6 on 51943] FSNamesystem.audit (FSNamesystem.java:logAuditMessage(7598)) - allowed=true    ugi=root (auth:SIMPLE)    ip=/127.0.0.1    cmd=listStatus    src=/    dst=null    perm=null    proto=rpc
2018-09-18 06:01:53,022 ERROR [core-4] writer.StorageWriter (StorageWriter.java:logError(457)) - StorageWriter[3]: Iteration[3].Error.
java.lang.IllegalArgumentException: Offset 0 is beyond the last offset 0 of the segment.
    at com.google.common.base.Preconditions.checkArgument(Preconditions.java:323)
    at io.pravega.segmentstore.storage.rolling.RollingStorage.read(RollingStorage.java:151)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$read$8(AsyncStorageWrapper.java:135)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.execute(AsyncStorageWrapper.java:221)
    at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$supplyAsync$13(AsyncStorageWrapper.java:181)
```
The test continues trying to read the `SegmentProperties` of segments in HDFS until the `120s` timeout for `waitForSegmentsInStorage` is exhausted and the test fails.

**Problem location**
`HDFSIntegrationTest.testEndToEndWithFencing`

**Suggestions for an improvement**
Make the test to pass consistently.

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**  
- Simplified the `StreamSegmentStoreTestBase.testEndToEndWithFencing` to do fewer fencing tests (4, down from 5) and to not do segment mergers or seals. 
- Added the ability to assign an Instance Id to a Segment Store instance, which will be output during logging. 
- Fixed two bugs in `RollingStorage` which could leave upstream code in a bad state.

**Purpose of the change**  
Fixes #2933.

**What the code does**  
- Instance Id is useful if multiple Segment Store instances run in the same process or output to the same log files. This will be included in the Thread names so that log output can be properly associated with an instance.
    - This is defaulted to empty string, which means for normal, production code, it will not add anything. It's just set for integration tests for now.
- The `testEndToEndWithFencing` was too complex and was taking too much time. It has been simplified with fewer operations that will execute.
- `RollingStorage`
    - Removed a parameter check which was violating the Javadoc contract. It would disallow reading 0 bytes from offset 0 of an empty file, when normally that should return an empty result.
    - Handling a concurrent rollover scenario, in case a different instance of the `RollingStorage` class executes a rollover on the Segment at or around the same time as this instance. This will force-refresh the handle and either move on or re-throw the original exception, depending on the case.

**How to verify it**  
`testEndToEndWithFencing` must pass consistently.

### CODE REVIEW

**Comment by RaulGracia** [comment]
By the way, I thought that this PR was related to #2933.

**Comment by fpj** [comment]
Build failure is a checkstyle one:

```
> Task :segmentstore:server:checkstyleTest
[ant:checkstyle] [ERROR] /home/travis/build/pravega/pravega/segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java:12:8: Unused import - com.google.common.collect.Iterators. [UnusedImports]
```

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java [hunk 1]
```diff
@@ -261,22 +261,17 @@ public void testEndToEndWithFencing() throws Exception {
             val segmentNames = createSegments(context.getActiveStore());
             val segmentsAndTransactions = new ArrayList<String>(segmentNames);
             log.info("Created Segments: {}.", String.join(", ", segmentNames));
-            val transactionsBySegment = createTransactions(segmentNames, context.getActiveStore());
```
[the remark this answers is not in the document]
Comment (andreipaduroiu) [reply segmentstore/server/src/test/java/io/pravega/segmentstore/server/store/StreamSegmentStoreTestBase.java]: The problems persisted even when removing these. It just happened that it got stuck on a transaction segment, which outnumbered the non-transaction segments.

**Comment by RaulGracia** [comment]
Executed 3 Jenkins builds without seen any failure in `testEndToEndWithFencing`.

**Comment by RaulGracia** [comment]
Executed system tests on this branch and all passed.

**Pull request merged**